File: plugins/wasm-go/pkg/matcher/rule_matcher.go
// Copyright (c) 2022 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package matcher

import (
	"errors"
	"fmt"
	"sort"
	"strings"

	"github.com/higress-group/proxy-wasm-go-sdk/proxywasm"
	"github.com/higress-group/proxy-wasm-go-sdk/proxywasm/types"
	"github.com/tidwall/gjson"

	"github.com/alibaba/higress/plugins/wasm-go/pkg/iface"
	"github.com/alibaba/higress/plugins/wasm-go/pkg/log"
)

type Category int

const (
	Route Category = iota
	Host
	Service
	RoutePrefix
	RouteAndService
)

type MatchType int

const (
	Prefix MatchType = iota
	Exact
	Suffix
)

const (
	RULES_KEY              = "_rules_"
	MATCH_ROUTE_KEY        = "_match_route_"
	MATCH_DOMAIN_KEY       = "_match_domain_"
	MATCH_SERVICE_KEY      = "_match_service_"
	MATCH_ROUTE_PREFIX_KEY = "_match_route_prefix_"
)

type HostMatcher struct {
	matchType MatchType
	host      string
}

type RuleConfig[PluginConfig any] struct {
	category     Category
	routes       map[string]struct{}
	services     map[string]struct{}
	routePrefixs map[string]struct{}
	hosts        []HostMatcher
	config       PluginConfig
}

// GenerateHashKey generates a hash key for the rule config based on matching conditions
func (r *RuleConfig[PluginConfig]) GenerateHashKey() string {
	var keyParts []string

	// Add category
	keyParts = append(keyParts, fmt.Sprintf("cat:%d", r.category))

	// Add routes (sorted for stability)
	if len(r.routes) > 0 {
		var routes []string
		for route := range r.routes {
			routes = append(routes, route)
		}
		sort.Strings(routes)
		keyParts = append(keyParts, fmt.Sprintf("routes:%s", strings.Join(routes, ",")))
	}

	// Add services (sorted for stability)
	if len(r.services) > 0 {
		var services []string
		for service := range r.services {
			services = append(services, service)
		}
		sort.Strings(services)
		keyParts = append(keyParts, fmt.Sprintf("services:%s", strings.Join(services, ",")))
	}

	// Add route prefixes (sorted for stability)
	if len(r.routePrefixs) > 0 {
		var prefixes []string
		for prefix := range r.routePrefixs {
			prefixes = append(prefixes, prefix)
		}
		sort.Strings(prefixes)
		keyParts = append(keyParts, fmt.Sprintf("prefixes:%s", strings.Join(prefixes, ",")))
	}

	// Add hosts (already in slice order, but sort for consistency)
	if len(r.hosts) > 0 {
		var hosts []string
		for _, hostMatcher := range r.hosts {
			hosts = append(hosts, fmt.Sprintf("%d:%s", hostMatcher.matchType, hostMatcher.host))
		}
		sort.Strings(hosts)
		keyParts = append(keyParts, fmt.Sprintf("hosts:%s", strings.Join(hosts, ",")))
	}

	return strings.Join(keyParts, "|")
}

type RuleMatcher[PluginConfig any] struct {
	ruleConfig      []RuleConfig[PluginConfig]
	globalConfig    PluginConfig
	hasGlobalConfig bool
}

func (m RuleMatcher[PluginConfig]) GetMatchConfig() (*PluginConfig, error) {
	host, err := proxywasm.GetHttpRequestHeader(":authority")
	if err != nil {
		return nil, err
	}
	routeName, err := proxywasm.GetProperty([]string{"route_name"})
	if err != nil && err != types.ErrorStatusNotFound {
		return nil, err
	}
	serviceName, err := proxywasm.GetProperty([]string{"cluster_name"})
	if err != nil && err != types.ErrorStatusNotFound {
		return nil, err
	}
	for _, rule := range m.ruleConfig {
		// category == Host
		if rule.category == Host {
			if m.hostMatch(rule, host) {
				return &rule.config, nil
			}
		}
		// category == Route
		if rule.category == Route {
			if _, ok := rule.routes[string(routeName)]; ok {
				return &rule.config, nil
			}
		}
		// category == Service
		if rule.category == Service {
			if m.serviceMatch(rule, string(serviceName)) {
				return &rule.config, nil
			}
		}
		// category == RouteAndService
		if rule.category == RouteAndService {
			if _, ok := rule.routes[string(routeName)]; ok {
				if m.serviceMatch(rule, string(serviceName)) {
					return &rule.config, nil
				}
			}
		}
		// category == RoutePrefix
		if rule.category == RoutePrefix {
			for routePrefix := range rule.routePrefixs {
				if strings.HasPrefix(string(routeName), routePrefix) {
					return &rule.config, nil
				}
			}
		}
	}
	if m.hasGlobalConfig {
		return &m.globalConfig, nil
	}
	return nil, nil
}

func (m *RuleMatcher[PluginConfig]) ParseRuleConfig(context iface.PluginContext, config gjson.Result,
	parsePluginConfig func(gjson.Result, *PluginConfig) error,
	parseOverrideConfig func(gjson.Result, PluginConfig, *PluginConfig) error) error {
	var rules []gjson.Result
	obj := config.Map()
	keyCount := len(obj)
	if keyCount == 0 {
		// enable globally for empty config
		m.hasGlobalConfig = true
		return parsePluginConfig(config, &m.globalConfig)
	}
	if rulesJson, ok := obj[RULES_KEY]; ok {
		rules = rulesJson.Array()
		keyCount--
	}
	var pluginConfig PluginConfig
	var globalConfigError error
	if keyCount > 0 {
		err := parsePluginConfig(config, &pluginConfig)
		if err != nil {
			globalConfigError = err
		} else {
			m.globalConfig = pluginConfig
			m.hasGlobalConfig = true
		}
	}
	if len(rules) == 0 {
		if m.hasGlobalConfig {
			return nil
		}
		return fmt.Errorf("parse config failed, no valid rules; global config parse error:%v", globalConfigError)
	}
	// Check if rule level config isolation is enabled
	isRuleLevelIsolation := context.IsRuleLevelConfigIsolation()

	var successfulRules []RuleConfig[PluginConfig]
	var hasAnyRuleParseError bool

	for _, ruleJson := range rules {
		var (
			rule RuleConfig[PluginConfig]
			err  error
		)
		rule.routes = m.parseRouteMatchConfig(ruleJson)
		rule.hosts = m.parseHostMatchConfig(ruleJson)
		rule.services = m.parseServiceMatchConfig(ruleJson)
		rule.routePrefixs = m.parseRoutePrefixMatchConfig(ruleJson)
		hasRoute := len(rule.routes) != 0
		hasHosts := len(rule.hosts) != 0
		hasService := len(rule.services) != 0
		hasRoutePrefix := len(rule.routePrefixs) != 0
		if boolToInt(hasRoute)+boolToInt(hasService)+boolToInt(hasHosts)+boolToInt(hasRoutePrefix) == 0 {
			return errors.New("there is at least one of  '_match_route_', '_match_domain_', '_match_service_' and '_match_route_prefix_' can present in configuration.")
		}
		if hasRoute {
			rule.category = Route
			if hasService {
				rule.category = RouteAndService
			}
		} else if hasHosts {
			rule.category = Host
		} else if hasService {
			rule.category = Service
		} else {
			rule.category = RoutePrefix
		}

		// Try to parse the rule config
		if parseOverrideConfig != nil {
			err = parseOverrideConfig(ruleJson, m.globalConfig, &rule.config)
		} else {
			err = parsePluginConfig(ruleJson, &rule.config)
		}

		if err != nil {
			hasAnyRuleParseError = true
			if isRuleLevelIsolation {
				log.Warnf("parse rule config failed for rule %s: %v, trying to load from backup", ruleJson.Raw, err)
				// Try to load from backup
				if loadedRuleJson := m.loadRuleJsonFromBackup(context, rule); loadedRuleJson.Exists() {
					log.Infof("found backup rule config: %s", loadedRuleJson.Raw)
					// Try to parse the rule config from loaded JSON
					if parseOverrideConfig != nil {
						err = parseOverrideConfig(loadedRuleJson, m.globalConfig, &rule.config)
					} else {
						err = parsePluginConfig(loadedRuleJson, &rule.config)
					}
					if err == nil {
						successfulRules = append(successfulRules, rule)
						log.Infof("successfully loaded rule from backup: %s", ruleJson.Raw)
						continue
					}
					log.Errorf("failed to parse backup rule config: %v", err)
				}
				log.Errorf("failed to load rule from backup, skipping rule: %s", ruleJson.Raw)
				continue
			} else {
				return err
			}
		}

		// Store successful rule to backup if rule level isolation is enabled
		if isRuleLevelIsolation {
			log.Debugf("storing rule to backup: %s", ruleJson.Raw)
			if storeErr := m.storeRuleToBackup(context, ruleJson, rule); storeErr != nil {
				log.Errorf("failed to store rule to backup: %v, rule: %s", storeErr, ruleJson.Raw)
			} else {
				log.Debugf("successfully stored rule to backup: %s", ruleJson.Raw)
			}
		}

		successfulRules = append(successfulRules, rule)
	}

	m.ruleConfig = successfulRules

	// If no successful rules and rule level isolation is enabled, return error
	if isRuleLevelIsolation && len(successfulRules) == 0 && hasAnyRuleParseError {
		return fmt.Errorf("all rules failed to parse and no previous successful config available")
	} else if !isRuleLevelIsolation && len(successfulRules) == 0 {
		return fmt.Errorf("no valid rules parsed")
	}
	return nil
}

func (m RuleMatcher[PluginConfig]) parseRouteMatchConfig(config gjson.Result) map[string]struct{} {
	keys := config.Get(MATCH_ROUTE_KEY).Array()
	routes := make(map[string]struct{})
	for _, item := range keys {
		routeName := item.String()
		if routeName != "" {
			routes[routeName] = struct{}{}
		}
	}
	return routes
}

func (m RuleMatcher[PluginConfig]) parseRoutePrefixMatchConfig(config gjson.Result) map[string]struct{} {
	keys := config.Get(MATCH_ROUTE_PREFIX_KEY).Array()
	routePrefixs := make(map[string]struct{})
	for _, item := range keys {
		routePrefix := item.String()
		if routePrefix != "" {
			routePrefixs[routePrefix] = struct{}{}
		}
	}
	return routePrefixs
}

func (m RuleMatcher[PluginConfig]) parseServiceMatchConfig(config gjson.Result) map[string]struct{} {
	keys := config.Get(MATCH_SERVICE_KEY).Array()
	clusters := make(map[string]struct{})
	for _, item := range keys {
		clusterName := item.String()
		if clusterName != "" {
			clusters[clusterName] = struct{}{}
		}
	}
	return clusters
}

func (m RuleMatcher[PluginConfig]) parseHostMatchConfig(config gjson.Result) []HostMatcher {
	keys := config.Get(MATCH_DOMAIN_KEY).Array()
	var hostMatchers []HostMatcher
	for _, item := range keys {
		host := item.String()
		var hostMatcher HostMatcher
		if strings.HasPrefix(host, "*") {
			hostMatcher.matchType = Suffix
			hostMatcher.host = host[1:]
		} else if strings.HasSuffix(host, "*") {
			hostMatcher.matchType = Prefix
			hostMatcher.host = host[:len(host)-1]
		} else {
			hostMatcher.matchType = Exact
			hostMatcher.host = host
		}
		hostMatchers = append(hostMatchers, hostMatcher)
	}
	return hostMatchers
}

func stripPortFromHost(reqHost string) string {
	// Port removing code is inspired by
	// https://github.com/envoyproxy/envoy/blob/v1.17.0/source/common/http/header_utility.cc#L219
	portStart := strings.LastIndexByte(reqHost, ':')
	if portStart != -1 {
		// According to RFC3986 v6 address is always enclosed in "[]".
		// section 3.2.2.
		v6EndIndex := strings.LastIndexByte(reqHost, ']')
		if v6EndIndex == -1 || v6EndIndex < portStart {
			if portStart+1 <= len(reqHost) {
				return reqHost[:portStart]
			}
		}
	}
	return reqHost
}

func (m RuleMatcher[PluginConfig]) hostMatch(rule RuleConfig[PluginConfig], reqHost string) bool {
	reqHost = stripPortFromHost(reqHost)
	for _, hostMatch := range rule.hosts {
		switch hostMatch.matchType {
		case Suffix:
			if strings.HasSuffix(reqHost, hostMatch.host) {
				return true
			}
		case Prefix:
			if strings.HasPrefix(reqHost, hostMatch.host) {
				return true
			}
		case Exact:
			if reqHost == hostMatch.host {
				return true
			}
		default:
			return false
		}
	}
	return false
}

func (m RuleMatcher[PluginConfig]) serviceMatch(rule RuleConfig[PluginConfig], serviceName string) bool {
	parts := strings.Split(serviceName, "|")
	if len(parts) != 4 {
		return false
	}
	port := parts[1]
	fqdn := parts[3]
	for configServiceName := range rule.services {
		colonIndex := strings.LastIndexByte(configServiceName, ':')
		if colonIndex != -1 && fqdn == string(configServiceName[:colonIndex]) && port == string(configServiceName[colonIndex+1:]) {
			return true
		} else if fqdn == string(configServiceName) {
			return true
		}
	}
	return false
}

var gRuleBackupStore map[string]string

func init() {
	gRuleBackupStore = make(map[string]string)
}

// storeRuleToSharedMemory stores a successful rule configuration to backup store
func (m *RuleMatcher[PluginConfig]) storeRuleToBackup(context iface.PluginContext, ruleJson gjson.Result, rule RuleConfig[PluginConfig]) error {
	hashKey := rule.GenerateHashKey()
	gRuleBackupStore[hashKey] = ruleJson.Raw
	log.Infof("store rule to backup, key[%s]", hashKey)
	return nil
}

// loadRuleJsonFromSharedMemory loads a rule JSON from backup store
func (m *RuleMatcher[PluginConfig]) loadRuleJsonFromBackup(context iface.PluginContext, rule RuleConfig[PluginConfig]) gjson.Result {
	hashKey := rule.GenerateHashKey()
	data, ok := gRuleBackupStore[hashKey]
	if !ok || data == "" {
		log.Infof("load rule from backup failed, key[%s]", hashKey)
		return gjson.Result{}
	}
	log.Infof("load rule from backup success, key[%s]", hashKey)
	return gjson.Parse(data)
}

File: plugins/wasm-assemblyscript/assembly/rule_matcher.ts
import { getRequestHost } from "./request_wrapper";
import {
  get_property,
  LogLevelValues,
  log,
  WasmResultValues,
} from "@higress/proxy-wasm-assemblyscript-sdk/assembly";
import { JSON } from "assemblyscript-json/assembly";

enum Category {
  Route,
  Host,
  RoutePrefix,
  Service
}

enum MatchType {
  Prefix,
  Exact,
  Suffix,
}

const RULES_KEY: string = "_rules_";
const MATCH_ROUTE_KEY: string = "_match_route_";
const MATCH_DOMAIN_KEY: string = "_match_domain_";
const MATCH_SERVICE_KEY: string = "_match_service_";
const MATCH_ROUTE_PREFIX_KEY: string = "_match_route_prefix_"

class HostMatcher {
  matchType: MatchType;
  host: string;

  constructor(matchType: MatchType, host: string) {
    this.matchType = matchType;
    this.host = host;
  }
}

class RuleConfig<PluginConfig> {
  category:      Category;
  routes!:       Map<string, boolean>;
  services!:     Map<string, boolean>;
  routePrefixs!: Map<string, boolean>;
  hosts!:        Array<HostMatcher>;
  config:        PluginConfig | null;

  constructor() {
    this.category = Category.Route;
    this.config = null;
  }
}

export class ParseResult<PluginConfig> {
  pluginConfig: PluginConfig | null;
  success: boolean;
  constructor(pluginConfig: PluginConfig | null, success: boolean) {
    this.pluginConfig = pluginConfig;
    this.success = success;
  }
}

export class RuleMatcher<PluginConfig> {
  ruleConfig: Array<RuleConfig<PluginConfig>>;
  globalConfig: PluginConfig | null;
  hasGlobalConfig: boolean;

  constructor() {
    this.ruleConfig = new Array<RuleConfig<PluginConfig>>();
    this.globalConfig = null;
    this.hasGlobalConfig = false;
  }

  getMatchConfig(): ParseResult<PluginConfig> {
    const host = getRequestHost();
    if (host == "") {
      return new ParseResult<PluginConfig>(null, false);
    }
    let result = get_property("route_name");
    if (result.status != WasmResultValues.Ok && result.status != WasmResultValues.NotFound) {
      return new ParseResult<PluginConfig>(null, false);
    }
    const routeName = String.UTF8.decode(result.returnValue);

    result = get_property("cluster_name");
    if (result.status != WasmResultValues.Ok && result.status != WasmResultValues.NotFound) {
      return new ParseResult<PluginConfig>(null, false);
    }
    const serviceName = String.UTF8.decode(result.returnValue);

    for (let i = 0; i < this.ruleConfig.length; i++) {
      const rule = this.ruleConfig[i];
      // category == Host
      if (rule.category == Category.Host) {
        if (this.hostMatch(rule, host)) {
          log(LogLevelValues.debug, "getMatchConfig: match host " + host);
          return new ParseResult<PluginConfig>(rule.config, true);
        }
      }
      // category == Route
      if (rule.category == Category.Route) {
        if (rule.routes.has(routeName)) {
          log(LogLevelValues.debug, "getMatchConfig: match route " + routeName);
          return new ParseResult<PluginConfig>(rule.config, true);
        }
      }
      // category == RoutePrefix
      if (rule.category == Category.RoutePrefix) {
        for (let i = 0; i < rule.routePrefixs.keys().length; i++) {
          const routePrefix = rule.routePrefixs.keys()[i];
          if (routeName.startsWith(routePrefix)) {
            return new ParseResult<PluginConfig>(rule.config, true);
          }
        }
      }
      // category == Cluster
      if (this.serviceMatch(rule, serviceName)) {
        return new ParseResult<PluginConfig>(rule.config, true);
      }
    }

    if (this.hasGlobalConfig) {
      return new ParseResult<PluginConfig>(this.globalConfig, true);
    }
    return new ParseResult<PluginConfig>(null, false);
  }

  parseRuleConfig(
    config: JSON.Obj,
    parsePluginConfig: (json: JSON.Obj) => ParseResult<PluginConfig>
  ): boolean {
    const obj = config;
    let keyCount = obj.keys.length;
    if (keyCount == 0) {
      this.hasGlobalConfig = true;
      const parseResult = parsePluginConfig(config);
      if (parseResult.success) {
        this.globalConfig = parseResult.pluginConfig;
        return true;
      } else {
        return false;
      }
    }

    let rules: JSON.Arr | null = null;
    if (obj.has(RULES_KEY)) {
      rules = obj.getArr(RULES_KEY);
      keyCount--;
    }

    if (keyCount > 0) {
      const parseResult = parsePluginConfig(config);
      if (parseResult.success) {
        this.globalConfig = parseResult.pluginConfig;
        this.hasGlobalConfig = true;
      }
    }

    if (!rules) {
      if (this.hasGlobalConfig) {
        return true;
      }
      log(LogLevelValues.error, "parse config failed, no valid rules; global config parse error");
      return false;
    }

    const rulesArray = rules.valueOf();
    for (let i = 0; i < rulesArray.length; i++) {
      if (!rulesArray[i].isObj) {
        log(LogLevelValues.error, "parse rule failed, rules must be an array of objects");
        continue;
      }
      const ruleJson = changetype<JSON.Obj>(rulesArray[i]);
      const rule = new RuleConfig<PluginConfig>();
      const parseResult = parsePluginConfig(ruleJson);
      if (parseResult.success) {
        rule.config = parseResult.pluginConfig;
      } else {
        return false;
      }

      rule.routes = this.parseRouteMatchConfig(ruleJson);
      rule.hosts = this.parseHostMatchConfig(ruleJson);
      rule.services = this.parseServiceMatchConfig(ruleJson);
      rule.routePrefixs = this.parseRoutePrefixMatchConfig(ruleJson);

      const noRoute = rule.routes.size == 0;
      const noHosts = rule.hosts.length == 0;
      const noServices = rule.services.size == 0;
      const noRoutePrefixs = rule.routePrefixs.size == 0;

      if ((boolToInt(noRoute) + boolToInt(noHosts) + boolToInt(noServices) + boolToInt(noRoutePrefixs)) != 3) {
        log(LogLevelValues.error, "there is only one of  '_match_route_', '_match_domain_', '_match_service_' and '_match_route_prefix_' can present in configuration.");
        return false;
      }
      if (!noRoute) {
        rule.category = Category.Route;
      } else if (!noHosts) {
        rule.category = Category.Host;
      } else if (!noServices) {
        rule.category = Category.Service;
      } else {
        rule.category = Category.RoutePrefix;
      }
      this.ruleConfig.push(rule);
    }
    return true;
  }

  parseRouteMatchConfig(config: JSON.Obj): Map<string, boolean> {
    const keys = config.getArr(MATCH_ROUTE_KEY);
    const routes = new Map<string, boolean>();
    if (keys) {
      const array = keys.valueOf();
      for (let i = 0; i < array.length; i++) {
        const key = array[i].toString();
        if (key != "") {
          routes.set(key, true);
        }
      }
    }
    return routes;
  }

  parseRoutePrefixMatchConfig(config: JSON.Obj): Map<string, boolean> {
    const keys = config.getArr(MATCH_ROUTE_PREFIX_KEY);
    const routePrefixs = new Map<string, boolean>();
    if (keys) {
      const array = keys.valueOf();
      for (let i = 0; i < array.length; i++) {
        const key = array[i].toString();
        if (key != "") {
          routePrefixs.set(key, true);
        }
      }
    }
    return routePrefixs;
  }

  parseServiceMatchConfig(config: JSON.Obj): Map<string, boolean> {
    const keys = config.getArr(MATCH_SERVICE_KEY);
    const clusters = new Map<string, boolean>();
    if (keys) {
      const array = keys.valueOf();
      for (let i = 0; i < array.length; i++) {
        const key = array[i].toString();
        if (key != "") {
          clusters.set(key, true);
        }
      }
    }
    return clusters;
  }

  parseHostMatchConfig(config: JSON.Obj): Array<HostMatcher> {
    const hostMatchers = new Array<HostMatcher>();
    const keys = config.getArr(MATCH_DOMAIN_KEY);
    if (keys !== null) {
      const array = keys.valueOf();
      for (let i = 0; i < array.length; i++) {
        const item = array[i].toString(); // Assuming the array has string elements
        let hostMatcher: HostMatcher;
        if (item.startsWith("*")) {
          hostMatcher = new HostMatcher(MatchType.Suffix, item.substr(1));
        } else if (item.endsWith("*")) {
          hostMatcher = new HostMatcher(
            MatchType.Prefix,
            item.substr(0, item.length - 1)
          );
        } else {
          hostMatcher = new HostMatcher(MatchType.Exact, item);
        }
        hostMatchers.push(hostMatcher);
      }
    }
    return hostMatchers;
  }

  stripPortFromHost(reqHost: string): string {
    // Port removing code is inspired by
    // https://github.com/envoyproxy/envoy/blob/v1.17.0/source/common/http/header_utility.cc#L219
    let portStart: i32 = reqHost.lastIndexOf(":");
    if (portStart != -1) {
      // According to RFC3986 v6 address is always enclosed in "[]".
      // section 3.2.2.
      let v6EndIndex: i32 = reqHost.lastIndexOf("]");
      if (v6EndIndex == -1 || v6EndIndex < portStart) {
        if (portStart + 1 <= reqHost.length) {
          return reqHost.substring(0, portStart);
        }
      }
    }
    return reqHost;
  }

  hostMatch(rule: RuleConfig<PluginConfig>, reqHost: string): boolean {
    reqHost = this.stripPortFromHost(reqHost);
    for (let i = 0; i < rule.hosts.length; i++) {
      let hostMatch = rule.hosts[i];
      switch (hostMatch.matchType) {
        case MatchType.Suffix:
          if (reqHost.endsWith(hostMatch.host)) {
            return true;
          }
          break;
        case MatchType.Prefix:
          if (reqHost.startsWith(hostMatch.host)) {
            return true;
          }
          break;
        case MatchType.Exact:
          if (reqHost == hostMatch.host) {
            return true;
          }
          break;
        default:
          return false;
      }
    }
    return false;
  }

  serviceMatch(rule: RuleConfig<PluginConfig>, serviceName: string): boolean {
    const parts = serviceName.split('|');
    if (parts.length != 4) {
      return false;
    }
    const port = parts[1];
    const fqdn = parts[3];
    for (let i = 0; i < rule.services.keys().length; i++) {
      let configServiceName = rule.services.keys()[i];
      let colonIndex = configServiceName.lastIndexOf(':');
      if (colonIndex != -1) {
          let configFQDN = configServiceName.slice(0, colonIndex);
          let configPort = configServiceName.slice(colonIndex + 1);
          if (fqdn == configFQDN && port == configPort) return true;
      } else if (fqdn == configServiceName) {
          return true;
      }
    }
    return false;
  }
}

function boolToInt(value: boolean): i32 {
  return value ? 1 : 0;
}